botocore/awsrequest.py

```diff
@@ -534,6 +557,84 @@ def reset_stream(self):
             raise UnseekableStreamError(stream_object=self.body)
 
 
+class AWSRequestCompressor:
+    """A class that can compress the body of an ``AWSRequest``."""
+
+    def compress(self, body, encodings, headers):
+        """Compresses the reqeust body using the specified encodings.
+
+        Set or append the `Content-Encoding` header with the matched encoding
+        if not present.
+        """
+        for encoding in encodings:
+            encoder = SUPPORTED_COMPRESSION_ENCODINGS.get(encoding)
+            if encoder is not None:
+                if 'Content-Encoding' not in headers:
+                    headers['Content-Encoding'] = encoding
+                elif encoding not in headers['Content-Encoding']:
+                    headers.replace_header(
+                        'Content-Encoding',
+                        f'{headers["Content-Encoding"]},{encoding}',
+                    )
+                return self._compress_body(body, encoder)
+            else:
+                logger.debug('Unsupported compression encoding: %s' % encoding)
+        return body
+
+    def _compress_body(self, body, encoder):
+        if isinstance(body, str):
+            return encoder(body.encode('utf-8'))
+        elif isinstance(body, (bytes, bytearray)):
+            return encoder(body)
+        elif hasattr(body, 'read'):
+            return encoder(b''.join(self._read_body(body)))
+        raise TypeError('Unable to compress body of type: %s' % type(body))
+
+    def _read_body(self, body):
+        while True:
+            # urllib3 and by extension the http stdlib uses a chunksize of 8192
+            # so we do the same here.
+            chunk = body.read(8192)
+            if not chunk:
+                break
+            if isinstance(chunk, str):
+                chunk = chunk.encode('utf-8')
+            yield chunk
+        if hasattr(body, 'seek'):
+            body.seek(0)
+
+    def should_compress_request(self, config, body, operation_model):
+        """Check if an AWSRequest object's body should be compressed."""
+        # compression must be explicitly disabled
+        if (
+            botocore.utils.ensure_boolean(config.disable_request_compression)
+            is not True
+        ):
```

Comment:
It wasn't previously since I intentionally chose not to coerce these settings in the config provider, but that's changed and as a result the `ensure_boolean` call is redundant here and will be removed in a future commit.